b2g/components/RemoteControlService.jsm

```diff
@@ -381,84 +407,129 @@ this.RemoteControlService = {
-              }
-
-              if (!allowed && self._port == -1) {
-                DEBUG && debug("Throw away TLSServerSocket with bad port.");
-                temp.close();
-                continue;
-              }
+    let self = this;
+    this._getCertificate().then(function(aCertificate){
+      // Store server certificate to synthesize PIN for JPAKE round2
+      self._certFingerprint = aCertificate.sha256Fingerprint;
 
-              socket = temp;
-              break;
-            }
+      if (!self._createSocket(aCertificate)) {
+        aReject("Start TLSSocketServer fail");
+        return;
+      }
 
-            if (!socket) {
-              throw new Error("No socket server available. Are there no available ports?");
-            }
+      let registrationListener = {
+        onServiceRegistered: function() {
+          DEBUG && debug ("Register service on port: " + self._port + " successfully");
+          self._serverStatus = SERVER_STATUS.STARTED;
+          aResolve();
+        },
+        onRegistrationFailed: function() {
+          DEBUG && debug ("Fail to register service");
+          aReject("Fail to register service");
+        },
+      };
 
-            DEBUG && debug("Listen on port " + socket.port + ", " + MAX_CLIENT_CONNECTIONS + " pending connections");
+      self._registerService(registrationListener);
+    }).catch(function(aError) {
+      aReject(aError);
+      return;
+    });
+  },
 
-            socket.serverCert = cert;
-            // Set session cache and tickets to false here.
-            // Cache disconnects fennect addon when addon sends message
-            // Tickets crashes b2g when fennec addon connects
-            socket.setSessionCache(false);
-            socket.setSessionTickets(false);
-            socket.setRequestClientCertificate(Ci.nsITLSServerSocket.REQUEST_NEVER);
+  // Get certificate without validation.
+  _getCertificate: function() {
+    return new Promise(function(aResolve, aReject) {
+      // Start TLSSocketServer with self-signed certification
+      // If there is no module use PSM before, handleCert result is SEC_ERROR_NO_MODULE (0x805A1FC0).
+      // Get PSM here ensure certService.getOrCreateCert works properly.
+      Cc["@mozilla.org/psm;1"].getService(Ci.nsISupports);
 
-            socket.asyncListen(self);
-            self._port = socket.port;
-            self._serverSocket = socket;
-          } catch (e) {
-            DEBUG && debug("Could not start server on port " + self._port + ": " + e);
-            aReject("Start TLSSocketServer fail");
-            return;
+      // Try toget certificate from x509cert DB first.
+      // If the certificate doesn't exist, then create one.
+      try {
+        let certDB = Cc["@mozilla.org/security/x509certdb;1"]
+                     .getService(Ci.nsIX509CertDB);
+        let cert = certDB.findCertByNickname(SERVICE_CERT);
+        DEBUG && debug("Get the existing certificate");
+        aResolve(cert);
```

Comment:
you can try following style, which will avoid the ugly |#else| block.

#ifdef MOZ_WIDGET_GONK
  try {
    let cert = certDB.findCertByNickname(...);
    aResolve(cert);
    return;
  } catch (e) {
    // cert cannot be found, fall through to create new one.
  }
#endif

certService.getOrCreate(...);